catboost/R-package/src/catboostr.cpp

```diff
@@ -958,4 +962,97 @@ EXPORT_FUNCTION CatBoostEvaluateObjectImportances_R(
 EXPORT_FUNCTION CatBoostIsNullHandle_R(SEXP handleParam) {
     return ScalarLogical(!R_ExternalPtrAddr(handleParam));
 }
+
+EXPORT_FUNCTION CatBoostEvalMetrics_R(
+        SEXP modelParam,
+        SEXP poolParam,
+        SEXP metricsParam,
+        SEXP treeCountStartParam,
+        SEXP treeCountEndParam,
+        SEXP evalPeriodParam,
+        SEXP threadCountParam,
+        SEXP tmpDirParam,
+        SEXP resultDirParam) {
+
+    SEXP result = NULL;
+    size_t metricsCount;
+
+    R_API_BEGIN()
+    TFullModelHandle model = reinterpret_cast<TFullModelHandle>(R_ExternalPtrAddr(modelParam));
+    TPoolHandle pool = reinterpret_cast<TPoolHandle>(R_ExternalPtrAddr(poolParam));
+
+    size_t metricsParamLen = length(metricsParam);
+    TVector<TString> metricDescriptions;
+    TVector<NCatboostOptions::TLossDescription> metricLossDescriptions;
+    metricDescriptions.reserve(metricsParamLen);
+    metricLossDescriptions.reserve(metricsParamLen);
+    for (size_t i = 0; i < metricsParamLen; ++i) {
+      TString metricDescription = CHAR(asChar(VECTOR_ELT(metricsParam, i)));
+      metricDescriptions.push_back(metricDescription);
+      metricLossDescriptions.emplace_back(NCatboostOptions::ParseLossDescription(metricDescription));
+    }
+    auto metrics = CreateMetrics(metricLossDescriptions, model->GetDimensionsCount());
+
+    NPar::TLocalExecutor executor;
+    executor.RunAdditionalThreads(UpdateThreadCount(asInteger(threadCountParam)) - 1);
+
+    TMetricsPlotCalcer plotCalcer = CreateMetricCalcer(
+        *model,
+        asInteger(treeCountStartParam),
+        asInteger(treeCountEndParam),
+        asInteger(evalPeriodParam),
+        /*processedIterationsStep=*/50,
+        CHAR(asChar(tmpDirParam)),
+        metrics,
+        &executor
+    );
+
+    TRestorableFastRng64 rand(0);
+    auto processedDataProvider = CreateModelCompatibleProcessedDataProvider(
+        *pool,
+        metricLossDescriptions,
+        *model,
+        GetMonopolisticFreeCpuRam(),
+        &rand,
+        &executor
+    );
+
+    if (plotCalcer.HasAdditiveMetric()) {
+        plotCalcer.ProceedDataSetForAdditiveMetrics(processedDataProvider);
+    }
+    if (plotCalcer.HasNonAdditiveMetric()) {
+        while (!plotCalcer.AreAllIterationsProcessed()) {
+            plotCalcer.ProceedDataSetForNonAdditiveMetrics(processedDataProvider);
+            plotCalcer.FinishProceedDataSetForNonAdditiveMetrics();
+        }
+    }
+
+    TVector<TVector<double>> metricsScore = plotCalcer.GetMetricsScore();
+    plotCalcer.SaveResult(CHAR(asChar(resultDirParam)), /*metricsFile=*/"", /*saveMetrics*/ false, /*saveStats=*/true).ClearTempFiles();
+
+    auto metricsResult = CreateMetricsFromDescription(metricDescriptions, model->GetDimensionsCount());
+    metricsCount = metricsResult.ysize();
+
+    result = PROTECT(allocVector(VECSXP, metricsCount));
```

Comment:
I think if this allocation or the next ones fail it will cause a memory leak due to not destructing the C++ objects from the earlier code. Is the size known in advance from the input parameters? Could it be allocated earlier?